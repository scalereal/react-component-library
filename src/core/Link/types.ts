import React, { RefObject  } from 'react';
import { AnyStyledComponent } from 'styled-components';

export interface WithStyle {
    Style?: AnyStyledComponent;
}

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
export type HTMLPropsReact<T> = Omit<React.HTMLProps<T>, 'ref' | 'as'>;

export interface HTMLProps<T> extends HTMLPropsReact<T> {
    ref?: ((instance: T) => void) | RefObject<T>;
}
export interface WithThemeProp {
    // eslint-disable-next-line no-undef
    as?: keyof JSX.IntrinsicElements | React.ComponentType<any>;
}
export interface LinkProps extends HTMLProps<HTMLAnchorElement> {
    /** Link url */
    href: string;
}
