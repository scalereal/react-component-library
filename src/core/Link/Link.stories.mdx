import Link, { linkSizes } from './Link';
import { Preview, Story, Meta, Props, ArgsTable, Canvas } from '@storybook/addon-docs/blocks';
import { withKnobs, color, boolean, select } from '@storybook/addon-knobs';

<Meta
    title="Components/Core/Link"
    component={Link}
    argTypes={{
        linkColor: {
            control: 'color'
        },
        hoverColor: {
            control: 'color'
        },
        activeColor: {
            control: 'color'
        },
        linkSize: {
            control: {
                type: 'select',
                options: Object.values(linkSizes)
            }
        }
    }}
/>

# Link

Displays a Link

export const Template = ({ linkSize, linkColor, hoverColor, activeColor, ...args }) => {
    return (
        <Link href="#" linkSize={linkSize} linkColor={linkColor} hoverColor={hoverColor} activeColor={activeColor} {...args}>
            View Link
        </Link>
    );
};

<Preview>
    <Story
        name="Link"
        args={{
            linkSize: 'S'
        }}
    >
        {Template.bind({})}
    </Story>
</Preview>

### Color Variations

You can either change the theme or pass the colors as props.

<Preview withToolbar>
    <div style={{ display: 'flex', flexWrap: 'wrap', justifyContent: 'space-around' }}>
        <Link href="#" linkColor="blue" hoverColor="red">
            View Link
        </Link>
        <Link href="#" linkColor="red" hoverColor="blue">
            View Link
        </Link>
    </div>
</Preview>

### Sizes

There are five sizes available `XS` | `S` | `M` | `L` | `XL`.

<Preview withToolbar>
    <div style={{ display: 'flex', flexWrap: 'wrap', justifyContent: 'space-around' }}>
        <Link href="#" linkSize="XS" linkColor="blue">
            View Link
        </Link>
        <Link href="#" linkSize="S" linkColor="blue">
            View Link
        </Link>
        <Link href="#" linkSize="M" linkColor="blue">
            View Link
        </Link>
        <Link href="#" linkSize="L" linkColor="blue">
            View Link
        </Link>
        <Link href="#" linkSize="XL" linkColor="blue">
            View Link
        </Link>
    </div>
</Preview>

### Props

<Props story="Link" />
