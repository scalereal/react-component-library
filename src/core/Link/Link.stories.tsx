import { Meta, Story } from '@storybook/react/types-6-0'
import React from 'react'
import Link  from './Link'
import { LinkProps } from './types'

const linkSizes : LinkProps['linkSize'][] = ['XS', 'S', 'M', 'L', 'XL']

export default {
    title: 'Components/Core/Link',
    component: Link,
    argTypes: {
        linkColor: {
            control: 'color'
        },
        hoverColor: {
            control: 'color'
        },
        activeColor: {
            control: 'color'
        },
        linkSize:{
            control: {
                type: 'select',
                options: Object.values(linkSizes),
            },
        }
    }
} as Meta;

const Template:Story = ({linkSize,linkColor,hoverColor,activeColor,...args}) => {
    return (
    <Link href="#" linkSize={linkSize} linkColor={linkColor} hoverColor={hoverColor} activeColor={activeColor} {...args}>View Link</Link>
    )
}

export const Default = Template.bind({});
Default.args = {
    linkSize:'S'
}

