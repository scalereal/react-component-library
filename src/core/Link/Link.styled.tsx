import styled from "styled-components"
import linkTheme from "../../theme/core/link/index"
import { LinkProps } from './types';

export const LinkStyled = styled('a')<LinkProps>`
    color: ${({ theme, linkColor }) => (linkColor ? linkColor : theme.defaults.linkColor)};
    background-color: transparent;
    text-decoration: none;
    font-size: ${({ theme, linkSize }) => (theme.sizes[linkSize].fontSize)};
    &:active {
        color: ${({ theme, activeColor }) => (activeColor ? activeColor : theme.defaults.activeColor)};
    }

    &:hover {
        color: ${({ theme, hoverColor }) => (hoverColor ? hoverColor : theme.defaults.hoverColor)};
        text-decoration: underline;
        outline-width: 0;
        cursor: pointer;
    }
`;

LinkStyled.defaultProps = {
    theme:linkTheme
}