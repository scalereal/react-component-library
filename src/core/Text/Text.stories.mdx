import { Text, textSizes, textAlignments, textWeights } from './Text';
import { Preview, Story, Meta, Props, ArgsTable, Canvas } from '@storybook/addon-docs/blocks';
import { withKnobs, color, boolean, select } from '@storybook/addon-knobs';

<Meta
    title="Components/Core/Text"
    component={Text}
    argTypes={{
        hoverEffect: {
            control: { type: 'boolean' }
        },
        textColor: {
            control: 'color'
        },
        hoverColor: {
            control: 'color'
        },
        activeColor: {
            control: 'color'
        },
        textSize: {
            control: {
                type: 'select',
                options: Object.values(textSizes)
            }
        },
        textAlign: {
            control: {
                type: 'select',
                options: Object.values(textAlignments)
            }
        },
        textWeight: {
            control: {
                type: 'select',
                options: Object.values(textWeights)
            }
        }
    }}
/>

# Text

Displays a Text

export const Template = ({
    textColor,
    activeColor,
    hoverColor,
    lineThrough,
    uppercase,
    textSize,
    lineHeight,
    textWeight,
    textAlign,
    ...args
}) => {
    return (
        <Text
            textColor={textColor}
            activeColor={activeColor}
            hoverColor={hoverColor}
            lineThrough={lineThrough}
            uppercase={uppercase}
            textSize={textSize}
            lineHeight={lineHeight}
            textWeight={textWeight}
            textAlign={textAlign}
            {...args}
        >
            Lorem Ipsum is simply dummy text
        </Text>
    );
};

<Preview>
    <Story
        name="Text"
        args={{
            textSize: 'S'
        }}
    >
        {Template.bind({})}
    </Story>
</Preview>

### Color Variations

You can either change the theme or pass the colors as props.

<Preview withToolbar>
    <div style={{ display: 'flex', flexWrap: 'wrap', justifyContent: 'space-around' }}>
        <Text textColor="blue">Lorem Ipsum is simply dummy text</Text>
        <Text textColor="green">Lorem Ipsum is simply dummy text</Text>
    </div>
</Preview>

### Sizes

There are five sizes available `XS` | `S` | `M` | `L` | `XL`.

<Preview withToolbar>
    <div style={{ display: 'flex', flexWrap: 'wrap', justifyContent: 'space-around' }}>
        <Text textSize="XS">Lorem Ipsum is simply dummy text</Text>
        <Text textSize="S">Lorem Ipsum is simply dummy text</Text>
        <Text textSize="M">Lorem Ipsum is simply dummy text</Text>
        <Text textSize="L">Lorem Ipsum is simply dummy text</Text>
        <Text textSize="XL">Lorem Ipsum is simply dummy text</Text>
    </div>
</Preview>

### Props

<Props story="Text" />
