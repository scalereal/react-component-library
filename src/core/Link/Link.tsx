import React, { FC, PropsWithChildren } from 'react';
import Text from '../Text/Text';
import * as Styled from './Link.styled';
import { LinkProps } from './types';

export const Link: FC<LinkProps> = ({ href, children, ...props }:PropsWithChildren<LinkProps>) =>{
    const isValidStringOrNumber = (e: any) => typeof e === 'string' || typeof e === 'number';
        return (
            <Styled.LinkStyled href={href} {...props}>
                {React.Children.map(children, c => {
                    return isValidStringOrNumber(c) ? <Text>{c}</Text> : c;
                })}
            </Styled.LinkStyled>
        );
}
export default Link