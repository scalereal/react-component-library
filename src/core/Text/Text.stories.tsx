import { Meta, Story } from '@storybook/react/types-6-0';
import React from 'react';
import Text from "./Text"


export default {
    title: 'Components/Core/Text',
    component: Text,
    argTypes : {
        textColor: {
            control:'color'
        },
    }
} as Meta

const Template:Story = ({ textColor, lineThrough, uppercase, textSize, lineHeight, ...args}) => {
    return <Text textColor={textColor} lineThrough={lineThrough} uppercase={uppercase} textSize={textSize} lineHeight={lineHeight} {...args}>
    Lorem Ipsum is simply dummy text
    </Text>
}

export const Default = Template.bind({});
Default.args = {
    textColor:"#435465",
    textSize:'S'
}
