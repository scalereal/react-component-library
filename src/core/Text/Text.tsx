import React, { FC } from 'react'
import { WithStyle } from '../../utils'
import * as Styled from './Text.styled'
import { TextProps } from './types'

export const textSizes : TextProps['textSize'][] = ['XS', 'S', 'M', 'L', 'XL']
export const textAlignments : TextProps['textAlign'][] = ['left', 'right', 'center', 'justify', 'justify-all', 'start', 'end', 'match-parent', 'inherit', 'initial', 'unset']
export const textWeights : TextProps['textWeight'][] = ['light', 'regular', 'medium', 'strong', 'extraStrong']

const Text: FC<TextProps> & WithStyle = React.memo(
    React.forwardRef(({ children, textSize, textWeight, textAlign, ...restProps }, ref) => {
        return (
            <Styled.Text  textSize={textSize} textWeight={textWeight} textAlign={textAlign} ref={ref} {...restProps }>
                {children}
            </Styled.Text>
        );
    })
);

Text.defaultProps = {
    lineThrough: false,
    uppercase: false,
    textSize:'S'
};
Text.displayName = 'Text';
Text.Style = Styled.Text;

export default Text;
