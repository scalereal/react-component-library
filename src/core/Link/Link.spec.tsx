import { render } from '@testing-library/react';
import React from 'react';
import Text from '../Text/Text';
import Link  from './Link';

describe('Link', () => {
    it('should render with default props', () => {
        const { container } = render(<Link  href="#">Test</Link>);
        expect(container).toMatchSnapshot();
    });

    it('should render with component as child', () => {
        const { container } = render(
            <Link href="#">
                <Text textSize='S'>Hi i am Name</Text>
            </Link>
        );
        expect(container).toMatchSnapshot();
    });
});
