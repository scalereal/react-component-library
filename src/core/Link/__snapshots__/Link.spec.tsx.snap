// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should render with component as child 1`] = `
<div>
  <a
    class="sc-gsTCUz iyrXEY"
    href="#"
  >
    <span
      class="sc-bdfBwQ cJAVCI"
    >
      Hi i am Name
    </span>
  </a>
</div>
`;

exports[`Link should render with default props 1`] = `
<div>
  <a
    class="sc-gsTCUz iyrXEY"
    href="#"
  >
    <span
      class="sc-bdfBwQ cJAVCI"
    >
      Test
    </span>
  </a>
</div>
`;
